core/queryparser/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/ast/ASTServiceGraphPattern.java

```diff
@@ -84,6 +81,27 @@ public String getPatternString() {
 		return patternString;
 	}
 
+	private String getSourceString() {
+		Node theParent = getParentContainer(this);
+		String sourceString = null;
+		if (theParent instanceof ASTOperationContainer) {
+			sourceString = ((ASTOperationContainer) theParent).getSourceString();
+		} else if (theParent instanceof ASTUpdateSequence) {
+			sourceString = ((ASTUpdateSequence) theParent).getSourceString();
+		}
+
+		while (sourceString == null && theParent != null) {
+			theParent = theParent.jjtGetParent();
+			if (theParent == null) {
+				break;
+			}
+			if (theParent instanceof ASTUpdateSequence) {
+				sourceString = ((ASTUpdateSequence) theParent).getSourceString();
+			}
+		}
+		return sourceString;
```

Comment:
I was wondering if maybe there could be a null check here, so that any exception is thrown when looking for the source string instead of when using the resulting string. Maybe `return Objects.requireNonNull(sourceString);`

Unless of course there is a case where returning null from getSourceString() is acceptable.

Reply:
Thought of that for a while, but decided that if we did not found the `sourceString` at this point, something else was really messed up, so throwing a custom exception early, would not help much here since we do not have a resonable alternative to continue processing update further - a NPE will still be thrown in the code that uses it ...